src/Components/Components/src/Routing/Router.cs

```diff
@@ -155,12 +191,51 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path, bool isNavigationIntercepted)
+        {
+            // If this router instance does not provide an OnNavigateAsync parameter
+            // then we render the component associated with the route as per usual.
+            if (OnNavigateAsync == null)
+            {
+                Refresh(isNavigationIntercepted);
+                return;
+            }
+
+            // If we've already invoked a task and stored its CTS, then
+            // cancel the existing task.
+            if (_onNavigateCts != null && !_onNavigateCts.IsCancellationRequested)
+            {
+                _onNavigateCts.Cancel();
+            }
+            // Create a new cancellation token source for this instance
+            _onNavigateCts = new CancellationTokenSource();
+
+            var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
+            var task = OnNavigateAsync(navigateContext);
+
+            // Create a cancellation task based on the cancellation token
+            // associated with the current running task.
+            var cancellationTaskSource = new TaskCompletionSource();
+            navigateContext.CancellationToken.Register(() =>
+                cancellationTaskSource.TrySetCanceled(navigateContext.CancellationToken));
```

Comment:
Should avoid one extra capturing delegate.

```suggestion
                cancellationTaskSource.TrySetCanceled((CancellationToken)navigateContext.CancellationToken), navigateContext.CancellationToken);
```

Reply:
I ended up changing this to `SetResult` instead of `TrySetCancelled`. I think it makes more sense for a task that is observing the cancellation token to complete when the token is cancelled than to cancel itself.